File: dmoj.ca-downloaded-submissions/Misc/rgpc18p3.py
import sys
input = sys.stdin.readline

n, t = map(int, input().split())
cups = [0 for _ in range(n)]
dif = [0]*(n+1)
for _ in range(t):
    a, b, c = map(int, input().split())
    dif[a-1] += c
    dif[b] -= c

cups[0] = dif[0]
for i in range(1, n):
    cups[i] = dif[i] + cups[i-1]

l = int(input())

length, bar = 1, l
first, last = 0,0

while first < n:
    while first < n and cups[first] >= bar:
        first += 1
        last += 1

    if first == n:
        break

    sum = 0
    for i in range(last, first + 1):
        sum += cups[i]

    length += 1
    first += 1
    while sum <= l and first < n:
        length += 1
        first += 1
        sum += cups[first]
    if sum > l:
        bar = last

print(length - 1)

File: dmoj.ca-downloaded-submissions/Misc/rgpc18p3.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class MainClass {

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st = new StringTokenizer("");
	static String next() throws IOException {
		while (st == null || !st.hasMoreTokens())
			st = new StringTokenizer(br.readLine().trim());
		return st.nextToken();
	}
	
	static int nextInt() throws NumberFormatException, IOException {
		return Integer.parseInt(next());
	}
	
	public static void main(String[] args) throws IOException {
		int n = nextInt(), t = nextInt();
		int[] cups = new int[n+1];
		
		for(int i = 0; i < t; i++) {
			int a=nextInt()-1,b=nextInt(),c=nextInt();
			cups[a] += c; cups[b] -= c;
		}
		for(int i = 1; i < n; i++) {
			cups[i] = cups[i] + cups[i-1];
		}
		int l = nextInt();
		int length = 0;
		int sum = cups[0];
		int first = 0, last = 0;
		
		while(first < n) {
			while(first < n && sum <= l) {
				first++;
				sum += cups[first];
			}
			length = Math.max(length, first-last);
			sum -= cups[last];
			last++;
		}
		System.out.println(length);
	}
}